core/src/main/scala/akka/persistence/cassandra/journal/CassandraRecovery.scala

```diff
@@ -93,9 +102,17 @@ trait CassandraRecovery extends CassandraTagRecovery with TaggedPreparedStatemen
           extractor = Extractors.persistentRepr(eventDeserializer, serialization),
           // run the query on the journal dispatcher (not the queries dispatcher)
           dispatcher = sessionSettings.pluginDispatcher)
-        .map(p => queries.mapEvent(p.persistentRepr))
-        .runForeach(replayCallback)
-        .map(_ => ())
+        .map { p =>
+          println(s"# asyncReplayMessages mapEvent ${Thread.currentThread().getName}") // FIXME
+          queries.mapEvent(p.persistentRepr)
+        }
+        .map { p =>
+          println(s"# asyncReplayMessages replayCallback ${Thread.currentThread().getName}") // FIXME
+          replayCallback(p)
+        }
+        .withAttributes(ActorAttributes.dispatcher(sessionSettings.pluginDispatcher))
+        .runWith(Sink.ignore)
+        .map(_ => ())(akka.dispatch.ExecutionContexts.sameThreadExecutionContext)
     }
```

Comment:
I have only looked at the recovery usage of eventsByPersistenceId so far. With these changes it's all on the plugin dispatcher, which it wasn't before.

Wonder if the inner ActorAttributes.dispatcher in CassandraReadJournal will still create an unwanted async boundary, even though same dispatcher? I'll check.

I have to verify with Akka 2.6 also.

Reply:
> Wonder if the inner ActorAttributes.dispatcher in CassandraReadJournal will still create an unwanted async boundary, even though same dispatcher?

That's how it's documented anyway, so we would have to get rid of that if we are looking for optimal (and then maybe the attribute bugs will make that impossible).

I think I'll leave that part of the investigation to the 1.0 (master) version.

Reply:
The concern with the read side was _if_ the user runs their side on a different dispatcher and get two async boundaries, but I think I was wrong about that, since the operators with unspecified dispatcher would "inherit" the one the user selected, so it would only be the one between the plugin dispatcher and "the rest of the stream".